lms/djangoapps/program_enrollments/management/commands/link_program_enrollments.py

```diff
@@ -0,0 +1,134 @@
+        Returns:
+            dict mapping external user keys to lms usernames
+        """
+        result = {}
+        for user_item in user_items:
+            split_args = user_item.split(':')
+            if len(split_args) != 2:
+                message = (INCORRECT_PARAMETER_TPL).format(user_item)
+                raise CommandError(message)
+
+            external_user_key = split_args[0]
+            lms_username = split_args[1]
+            if external_user_key in result:
+                raise CommandError(DUPLICATE_KEY_TPL.format(external_user_key))
+
+            result[external_user_key] = lms_username
+        return result
+
+    def get_program_enrollments(self, program_uuid, external_student_keys):
+        """
+        Does a bulk read of ProgramEnrollments for a given program and list of external student keys
+        and returns a dict keyed by external student key
+        """
+        qs = ProgramEnrollment.bulk_read_by_student_key(program_uuid, external_student_keys)
+        qs = qs.prefetch_related('program_course_enrollments')
+        qs = qs.select_related('user')
+        return {
+            program_enrollment.external_user_key: program_enrollment
+            for program_enrollment in qs.iterator()
+        }
+
+    def get_lms_users(self, lms_usernames):
+        """
+        Does a bulk read of Users by username and returns a dict keyed by username
+        """
+        return {
+            user.username: user
+            for user in User.objects.filter(username__in=lms_usernames).iterator()
+        }
+
+    def link_program_enrollment(self, program_enrollment, user):
+        """
+        Attempts to link the given program enrollment to the given user
+        If the enrollment has any program course enrollments, enroll the user in those courses as well
+        """
+        if program_enrollment.user:
+            self.log_existing_user_message(program_enrollment, user)
+            return
+        logger.info(u'Linking external student key {} and user {}'.format(
+            program_enrollment.external_user_key,
+            user.username
+        ))
+        program_enrollment.user = user
+        program_enrollment.save()
+
+        for program_course_enrollment in program_enrollment.program_course_enrollments.all():
+            try:
+                program_course_enrollment.enroll(user)
+            except CourseEnrollmentException:
+                logger.warning(COURSE_ENROLLMENT_FAILURE_TPL.format(
+                    user.username,
+                    program_course_enrollment.course_key
+                ))
+
+    def log_existing_user_message(self, program_enrollment, user):
+        """
+        Logs an error message that the specified program enrollment is already linked to an lms user
+        """
+        existing_username = program_enrollment.user.username
+        external_student_key = program_enrollment.external_user_key
+        logger.warning(EXISTING_USER_TPL.format(
```

Comment:
Nit: could you use the parameter names of the template here?  It's hard to know what "target" or "a different" are supposed to map to without scrolling up to look at the template.